File: android/uhabits-core/src/main/java/org/isoron/uhabits/core/models/EntryList.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.core.models;

import androidx.annotation.*;

import org.apache.commons.lang3.builder.*;
import org.isoron.uhabits.core.utils.*;

import java.io.*;
import java.text.*;
import java.util.*;

import javax.annotation.concurrent.*;

import static org.isoron.uhabits.core.models.Entry.*;
import static org.isoron.uhabits.core.utils.StringUtils.*;

/**
 * The collection of {@link Entry}s belonging to a habit.
 */
@ThreadSafe
public class EntryList
{
    protected final Habit habit;

    protected ArrayList<Entry> list;

    public EntryList(Habit habit)
    {
        this.habit = habit;
        this.list = new ArrayList<>();
    }

    @NonNull
    static List<Entry> buildEntriesFromInterval(Entry[] original,
                                                ArrayList<Interval> intervals)
    {
        if (original.length == 0) throw new IllegalArgumentException();

        Timestamp today = DateUtils.getTodayWithOffset();
        Timestamp begin = original[0].getTimestamp();
        if (intervals.size() > 0) begin = Timestamp.oldest(begin, intervals.get(0).begin);

        int nDays = begin.daysUntil(today) + 1;
        List<Entry> entries = new ArrayList<>(nDays);
        for (int i = 0; i < nDays; i++)
            entries.add(new Entry(today.minus(i), UNKNOWN));

        for (Interval interval : intervals)
        {
            for (int i = 0; i < interval.length(); i++)
            {
                Timestamp date = interval.begin.plus(i);
                int offset = date.daysUntil(today);
                if (offset < 0) continue;
                entries.set(offset, new Entry(date, YES_AUTO));
            }
        }

        for (Entry e : original)
        {
            Timestamp date = e.getTimestamp();
            int offset = date.daysUntil(today);
            int value = e.getValue();
            int prevValue = entries.get(offset).getValue();
            if (prevValue < value)
                entries.set(offset, new Entry(date, value));
        }

        return entries;
    }

    /**
     * For non-daily habits, some manual entries generate many
     * automatic entries. For example, for weekly habits, each repetition generates
     * seven checkmarks. For twice-a-week habits, two repetitions that are close
     * enough together also generate seven checkmarks.
     * <p>
     * This group of generated entries is represented by an interval. This function
     * computes the list of intervals for a given list of original entries. It tries
     * to build the intervals as far away in the future as possible.
     */
    @NonNull
    static ArrayList<Interval> buildIntervals(@NonNull Frequency freq,
                                              @NonNull Entry[] entries)
    {
        ArrayList<Entry> filteredEntries = new ArrayList<>();
        for (Entry e : entries)
            if (e.getValue() == YES_MANUAL)
                filteredEntries.add(e);

        int num = freq.getNumerator();
        int den = freq.getDenominator();

        ArrayList<Interval> intervals = new ArrayList<>();
        for (int i = 0; i < filteredEntries.size() - num + 1; i++)
        {
            Entry first = filteredEntries.get(i);
            Entry last = filteredEntries.get(i + num - 1);

            long distance = first.getTimestamp().daysUntil(last.getTimestamp());
            if (distance >= den) continue;

            Timestamp begin = first.getTimestamp();
            Timestamp center = last.getTimestamp();
            Timestamp end = begin.plus(den - 1);
            intervals.add(new Interval(begin, center, end));
        }

        return intervals;
    }

    /**
     * Starting from the second newest interval, this function tries to slide the
     * intervals backwards into the past, so that gaps are eliminated and
     * streaks are maximized.
     */
    static void snapIntervalsTogether(@NonNull ArrayList<Interval> intervals)
    {
        int n = intervals.size();
        for (int i = n - 2; i >= 0; i--)
        {
            Interval curr = intervals.get(i);
            Interval next = intervals.get(i + 1);

            int gapNextToCurrent = next.begin.daysUntil(curr.end);
            int gapCenterToEnd = curr.center.daysUntil(curr.end);

            if (gapNextToCurrent >= 0)
            {
                int shift = Math.min(gapCenterToEnd, gapNextToCurrent + 1);
                intervals.set(i, new Interval(curr.begin.minus(shift),
                        curr.center,
                        curr.end.minus(shift)));
            }
        }
    }

    public void add(List<Entry> entries)
    {
        list.addAll(entries);
        Collections.sort(list,
                (c1, c2) -> c2.getTimestamp().compare(c1.getTimestamp()));

    }

    /**
     * Returns the values for all the checkmarks, since the oldest repetition of
     * the habit until today.
     * <p>
     * If there are no repetitions at all, returns an empty array. The values
     * are returned in an array containing one integer value for each day since
     * the first repetition of the habit until today. The first entry
     * corresponds to today, the second entry corresponds to yesterday, and so
     * on.
     *
     * @return values for the checkmarks in the interval
     */
    @NonNull
    public synchronized final int[] getAllValues()
    {
        Entry oldestOriginal = habit.getOriginalEntries().getOldest();
        if (oldestOriginal == null) return new int[0];

        Timestamp fromTimestamp = oldestOriginal.getTimestamp();
        Timestamp toTimestamp = DateUtils.getTodayWithOffset();

        return getValues(fromTimestamp, toTimestamp);
    }

    /**
     * Returns the list of checkmarks that fall within the given interval.
     * <p>
     * There is exactly one checkmark per day in the interval. The endpoints of
     * the interval are included. The list is ordered by timestamp (decreasing).
     * That is, the first checkmark corresponds to the newest timestamp, and the
     * last checkmark corresponds to the oldest timestamp.
     *
     * @param from timestamp of the beginning of the interval.
     * @param to   timestamp of the end of the interval.
     * @return the list of checkmarks within the interval.
     */
    @NonNull
    public List<Entry> getByInterval(Timestamp from,
                                     Timestamp to)
    {
        compute();

        Timestamp newestComputed = new Timestamp(0);
        Timestamp oldestComputed = new Timestamp(0).plus(1000000);

        Entry newest = getNewestComputed();
        Entry oldest = getOldestComputed();
        if (newest != null) newestComputed = newest.getTimestamp();
        if (oldest != null) oldestComputed = oldest.getTimestamp();

        List<Entry> filtered = new ArrayList<>(
                Math.max(0, oldestComputed.daysUntil(newestComputed) + 1));

        for (int i = 0; i <= from.daysUntil(to); i++)
        {
            Timestamp t = to.minus(i);
            if (t.isNewerThan(newestComputed) || t.isOlderThan(oldestComputed))
                filtered.add(new Entry(t, Entry.UNKNOWN));
            else
                filtered.add(list.get(t.daysUntil(newestComputed)));
        }

        return filtered;

    }

    /**
     * Returns the checkmark for today.
     *
     * @return checkmark for today
     */
    @Nullable
    public synchronized final Entry getToday()
    {
        compute();
        Timestamp today = DateUtils.getTodayWithOffset();
        return getByInterval(today, today).get(0);
    }

    /**
     * Returns the value of today's checkmark.
     *
     * @return value of today's checkmark
     */
    public synchronized int getTodayValue()
    {
        Entry today = getToday();
        if (today != null) return today.getValue();
        else return UNKNOWN;
    }

    public synchronized int getThisWeekValue(int firstWeekday)
    {
        return getThisIntervalValue(DateUtils.TruncateField.WEEK_NUMBER, firstWeekday);
    }

    public synchronized int getThisMonthValue()
    {
        return getThisIntervalValue(DateUtils.TruncateField.MONTH, Calendar.SATURDAY);
    }


    public synchronized int getThisQuarterValue()
    {
        return getThisIntervalValue(DateUtils.TruncateField.QUARTER, Calendar.SATURDAY);
    }


    public synchronized int getThisYearValue()
    {
        return getThisIntervalValue(DateUtils.TruncateField.YEAR, Calendar.SATURDAY);
    }

    private int getThisIntervalValue(DateUtils.TruncateField truncateField, int firstWeekday)
    {
        List<Entry> groups = habit.getComputedEntries().groupBy(truncateField, firstWeekday, 1);
        if (groups.isEmpty()) return 0;
        return groups.get(0).getValue();
    }

    /**
     * Returns the values of the checkmarks that fall inside a certain interval
     * of time.
     * <p>
     * The values are returned in an array containing one integer value for each
     * day of the interval. The first entry corresponds to the most recent day
     * in the interval. Each subsequent entry corresponds to one day older than
     * the previous entry. The boundaries of the time interval are included.
     *
     * @param from timestamp for the oldest checkmark
     * @param to   timestamp for the newest checkmark
     * @return values for the checkmarks inside the given interval
     */
    public final int[] getValues(Timestamp from, Timestamp to)
    {
        if (from.isNewerThan(to)) return new int[0];

        List<Entry> entries = getByInterval(from, to);
        int values[] = new int[entries.size()];

        int i = 0;
        for (Entry c : entries)
            values[i++] = c.getValue();

        return values;
    }

    /**
     * Marks as invalid every checkmark that has timestamp either equal or newer
     * than a given timestamp. These checkmarks will be recomputed at the next
     * time they are queried.
     *
     * @param timestamp the timestamp
     */
    public void invalidateNewerThan(Timestamp timestamp)
    {
        list.clear();
    }

    /**
     * Writes the entire list of checkmarks to the given writer, in CSV format.
     *
     * @param out the writer where the CSV will be output
     * @throws IOException in case write operations fail
     */
    public final void writeCSV(Writer out) throws IOException
    {
        int values[];

        synchronized (this)
        {
            compute();
            values = getAllValues();
        }

        Timestamp timestamp = DateUtils.getToday();
        SimpleDateFormat dateFormat = DateFormats.getCSVDateFormat();

        for (int value : values)
        {
            String date = dateFormat.format(timestamp.toJavaDate());
            out.write(String.format("%s,%d\n", date, value));
            timestamp = timestamp.minus(1);
        }
    }

    /**
     * Computes and stores one checkmark for each day, from the first habit
     * repetition to today. If this list is already computed, does nothing.
     */
    protected final synchronized void compute()
    {
        final Timestamp today = DateUtils.getTodayWithOffset();

        Entry newest = getNewestComputed();
        if (newest != null && newest.getTimestamp().equals(today)) return;
        invalidateNewerThan(Timestamp.ZERO);

        Entry oldestRep = habit.getOriginalEntries().getOldest();
        if (oldestRep == null) return;
        final Timestamp from = oldestRep.getTimestamp();

        if (from.isNewerThan(today)) return;

        Entry reps[] = habit
                .getOriginalEntries()
                .getByInterval(from, today)
                .toArray(new Entry[0]);

        if (habit.isNumerical()) computeNumerical(reps);
        else computeYesNo(reps);
    }

    @Nullable
    protected Entry getNewestComputed()
    {
        if (list.isEmpty()) return null;
        return list.get(0);
    }

    @Nullable
    protected Entry getOldestComputed()
    {
        if (list.isEmpty()) return null;
        return list.get(list.size() - 1);
    }

    private void computeNumerical(Entry[] original)
    {
        if (original.length == 0) return;

        Timestamp today = DateUtils.getTodayWithOffset();
        Timestamp begin = original[0].getTimestamp();

        int nDays = begin.daysUntil(today) + 1;
        List<Entry> computed = new ArrayList<>(nDays);
        for (int i = 0; i < nDays; i++)
            computed.add(new Entry(today.minus(i), 0));

        for (Entry e : original)
        {
            int offset = e.getTimestamp().daysUntil(today);
            computed.set(offset, new Entry(e.getTimestamp(), e.getValue()));
        }

        add(computed);
    }

    private void computeYesNo(Entry[] original)
    {
        ArrayList<Interval> intervals;
        intervals = buildIntervals(habit.getFrequency(), original);
        snapIntervalsTogether(intervals);
        add(buildEntriesFromInterval(original, intervals));
    }

    public List<Entry> getAll()
    {
        Entry oldest = habit.getOriginalEntries().getOldest();
        if (oldest == null) return new ArrayList<>();
        return getByInterval(oldest.getTimestamp(), DateUtils.getTodayWithOffset());
    }

    static final class Interval
    {
        final Timestamp begin;

        final Timestamp center;

        final Timestamp end;

        Interval(Timestamp begin, Timestamp center, Timestamp end)
        {
            this.begin = begin;
            this.center = center;
            this.end = end;
        }

        public int length()
        {
            return begin.daysUntil(end) + 1;
        }

        @Override
        public boolean equals(Object o)
        {
            if (this == o) return true;

            if (o == null || getClass() != o.getClass()) return false;

            Interval interval = (Interval) o;

            return new EqualsBuilder()
                    .append(begin, interval.begin)
                    .append(center, interval.center)
                    .append(end, interval.end)
                    .isEquals();
        }

        @Override
        public int hashCode()
        {
            return new HashCodeBuilder(17, 37)
                    .append(begin)
                    .append(center)
                    .append(end)
                    .toHashCode();
        }

        @Override
        public String toString()
        {
            return new ToStringBuilder(this, defaultToStringStyle())
                    .append("begin", begin)
                    .append("center", center)
                    .append("end", end)
                    .toString();
        }
    }

    @NonNull
    public List<Entry> groupBy(DateUtils.TruncateField field, int firstWeekday)
    {
        return groupBy(field, firstWeekday, 0);
    }


    @NonNull
    public List<Entry> groupBy(DateUtils.TruncateField field,
                               int firstWeekday,
                               int maxGroups)
    {
        List<Entry> checks = getAll();

        int count = 0;
        Timestamp[] truncatedTimestamps = new Timestamp[checks.size()];
        int[] values = new int[checks.size()];

        for (Entry rep : checks)
        {
            Timestamp tt = rep.getTimestamp().truncate(field, firstWeekday);
            if (count == 0 || !truncatedTimestamps[count - 1].equals(tt))
            {
                if (maxGroups > 0 && count >= maxGroups) break;
                truncatedTimestamps[count++] = tt;
            }

            if (habit.isNumerical())
                values[count - 1] += rep.getValue();
            else if (rep.getValue() == Entry.YES_MANUAL)
                values[count - 1] += 1000;

        }

        ArrayList<Entry> groupedEntries = new ArrayList<>();
        for (int i = 0; i < count; i++)
        {
            Entry rep = new Entry(truncatedTimestamps[i], values[i]);
            groupedEntries.add(rep);
        }

        return groupedEntries;
    }
}

File: android/uhabits-core/src/main/java/org/isoron/uhabits/core/models/EntryList.kt
/*
 * Copyright (C) 2016-2020 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.core.models

import org.isoron.uhabits.core.models.Entry.Companion.SKIP
import org.isoron.uhabits.core.models.Entry.Companion.UNKNOWN
import org.isoron.uhabits.core.models.Entry.Companion.YES_AUTO
import org.isoron.uhabits.core.models.Entry.Companion.YES_MANUAL
import org.isoron.uhabits.core.utils.DateUtils
import java.util.ArrayList
import java.util.Calendar
import javax.annotation.concurrent.ThreadSafe
import kotlin.collections.set
import kotlin.math.min

@ThreadSafe
open class EntryList {

    private val entriesByTimestamp: HashMap<Timestamp, Entry> = HashMap()

    /**
     * Returns the entry corresponding to the given timestamp. If no entry with such timestamp
     * has been previously added, returns Entry(timestamp, UNKNOWN).
     */
    @Synchronized
    open fun get(timestamp: Timestamp): Entry {
        return entriesByTimestamp[timestamp] ?: Entry(timestamp, UNKNOWN)
    }

    /**
     * Returns one entry for each day in the given interval. The first element corresponds to the
     * newest entry, and the last element corresponds to the oldest. The interval endpoints are
     * included.
     */
    @Synchronized
    open fun getByInterval(from: Timestamp, to: Timestamp): List<Entry> {
        val result = mutableListOf<Entry>()
        if (from.isNewerThan(to)) return result
        var current = to
        while (current >= from) {
            result.add(get(current))
            current = current.minus(1)
        }
        return result
    }

    /**
     * Adds the given entry to the list. If another entry with the same timestamp already exists,
     * replaces it.
     */
    @Synchronized
    open fun add(entry: Entry) {
        entriesByTimestamp[entry.timestamp] = entry
    }

    /**
     * Returns all entries whose values are known, sorted by timestamp. The first element
     * corresponds to the newest entry, and the last element corresponds to the oldest.
     */
    @Synchronized
    open fun getKnown(): List<Entry> {
        return entriesByTimestamp.values.sortedBy { it.timestamp }.reversed()
    }

    /**
     * Truncates the timestamps of all known entries, then aggregates their values. This function
     * is used to generate bar plots where each bar shows the number of repetitions in a given week,
     * month or year.
     *
     * For boolean habits, the value of the aggregated entry equals to the number of YES_MANUAL
     * entries. For numerical habits, the value is the total sum. The field [firstWeekday] is only
     * relevant when grouping by week.
     */
    @Synchronized
    open fun groupBy(
        original: List<Entry>,
        field: DateUtils.TruncateField,
        firstWeekday: Int,
        isNumerical: Boolean,
    ): List<Entry> {
        val truncated = original.map {
            Entry(it.timestamp.truncate(field, firstWeekday), it.value)
        }
        val timestamps = mutableListOf<Timestamp>()
        val values = mutableListOf<Int>()
        for (i in truncated.indices) {
            if (i == 0 || timestamps.last() != truncated[i].timestamp) {
                timestamps.add(truncated[i].timestamp)
                values.add(0)
            }
            if (isNumerical) {
                if (truncated[i].value > 0) {
                    values[values.lastIndex] += truncated[i].value
                }
            } else {
                if (truncated[i].value == YES_MANUAL) {
                    values[values.lastIndex] += 1000
                }
            }
        }
        return timestamps.indices.map { Entry(timestamps[it], values[it]) }
    }

    /**
     * Replaces all entries in this list by entries computed automatically from another list.
     *
     * For boolean habits, this function creates additional entries (with value YES_AUTO) according
     * to the frequency of the habit. For numerical habits, this function simply copies all entries.
     */
    @Synchronized
    open fun recomputeFrom(
        originalEntries: EntryList,
        frequency: Frequency,
        isNumerical: Boolean,
    ) {
        clear()
        val original = originalEntries.getKnown()
        if (isNumerical) {
            original.forEach { add(it) }
        } else {
            val intervals = buildIntervals(frequency, original)
            snapIntervalsTogether(intervals)
            val computed = buildEntriesFromInterval(original, intervals)
            computed.filter { it.value != UNKNOWN }.forEach { add(it) }
        }
    }

    /**
     * Removes all known entries.
     */
    @Synchronized
    open fun clear() {
        entriesByTimestamp.clear()
    }

    /**
     * Returns the total number of successful entries for each month, grouped by day of week.
     * <p>
     * The checkmarks are returned in a HashMap. The key is the timestamp for
     * the first day of the month, at midnight (00:00). The value is an integer
     * array with 7 entries. The first entry contains the total number of
     * successful checkmarks during the specified month that occurred on a Saturday. The
     * second entry corresponds to Sunday, and so on. If there are no
     * successful checkmarks during a certain month, the value is null.
     *
     * @return total number of checkmarks by month versus day of week
     */
    @Synchronized
    fun computeWeekdayFrequency(isNumerical: Boolean): HashMap<Timestamp, Array<Int>> {
        val entries = getKnown()
        val map = hashMapOf<Timestamp, Array<Int>>()
        for ((originalTimestamp, value) in entries) {
            val weekday = originalTimestamp.weekday
            val truncatedTimestamp = Timestamp(
                originalTimestamp.toCalendar().apply {
                    set(Calendar.DAY_OF_MONTH, 1)
                }.timeInMillis
            )

            var list = map[truncatedTimestamp]
            if (list == null) {
                list = arrayOf(0, 0, 0, 0, 0, 0, 0)
                map[truncatedTimestamp] = list
            }

            if (isNumerical) {
                list[weekday] += value
            } else if (value == YES_MANUAL) {
                list[weekday] += 1
            }
        }
        return map
    }

    /**
     * Returns the values of the entries that fall inside a certain interval of time. The values
     * are returned in an array containing one integer value for each day of the interval. The
     * first entry corresponds to the most recent day in the interval. Each subsequent entry
     * corresponds to one day older than the previous entry. The boundaries of the time interval
     * are included.
     */
    @Deprecated("")
    @Synchronized
    fun getValues(from: Timestamp, to: Timestamp): IntArray {
        if (from.isNewerThan(to)) throw IllegalArgumentException()
        val nDays = from.daysUntil(to) + 1
        val result = IntArray(nDays) { UNKNOWN }
        getKnown().filter { entry ->
            !entry.timestamp.isNewerThan(to) && !entry.timestamp.isOlderThan(from)
        }.forEach { entry ->
            val offset = entry.timestamp.daysUntil(to)
            result[offset] = entry.value
        }
        return result
    }

    @Deprecated("")
    @Synchronized
    fun getAllValues(): IntArray {
        val entries = getKnown()
        if (entries.isEmpty()) return IntArray(0)
        var (fromTimestamp, _) = entries.last()
        val toTimestamp = DateUtils.getTodayWithOffset()
        if (fromTimestamp.isNewerThan(toTimestamp)) fromTimestamp = toTimestamp
        return getValues(fromTimestamp, toTimestamp)
    }

    @Deprecated("")
    @Synchronized
    open fun getThisWeekValue(firstWeekday: Int, isNumerical: Boolean): Int {
        return getThisIntervalValue(
            truncateField = DateUtils.TruncateField.WEEK_NUMBER,
            firstWeekday = firstWeekday,
            isNumerical = isNumerical
        )
    }

    @Deprecated("")
    @Synchronized
    open fun getThisMonthValue(isNumerical: Boolean): Int {
        return getThisIntervalValue(
            truncateField = DateUtils.TruncateField.MONTH,
            firstWeekday = Calendar.SATURDAY,
            isNumerical = isNumerical
        )
    }

    @Deprecated("")
    @Synchronized
    open fun getThisQuarterValue(isNumerical: Boolean): Int {
        return getThisIntervalValue(
            truncateField = DateUtils.TruncateField.QUARTER,
            firstWeekday = Calendar.SATURDAY,
            isNumerical = isNumerical
        )
    }

    @Deprecated("")
    @Synchronized
    open fun getThisYearValue(isNumerical: Boolean): Int {
        return getThisIntervalValue(
            truncateField = DateUtils.TruncateField.YEAR,
            firstWeekday = Calendar.SATURDAY,
            isNumerical = isNumerical
        )
    }

    private fun getThisIntervalValue(
        truncateField: DateUtils.TruncateField,
        firstWeekday: Int,
        isNumerical: Boolean,
    ): Int {
        val groups: List<Entry> = groupBy(getKnown(), truncateField, firstWeekday, isNumerical)
        return if (groups.isEmpty()) 0 else groups[0].value
    }

    data class Interval(val begin: Timestamp, val center: Timestamp, val end: Timestamp) {
        val length: Int
            get() = begin.daysUntil(end) + 1
    }

    /**
     * Converts a list of intervals into a list of entries. Entries that fall outside of any
     * interval receive value UNKNOWN. Entries that fall within an interval but do not appear
     * in [original] receive value YES_AUTO. Entries provided in [original] are copied over.
     *
     * The intervals should be sorted by timestamp. The first element in the list should
     * correspond to the newest interval.
     */
    companion object {
        fun buildEntriesFromInterval(
            original: List<Entry>,
            intervals: List<Interval>,
        ): List<Entry> {
            val result = arrayListOf<Entry>()
            if (original.isEmpty()) return result

            var from = original[0].timestamp
            var to = original[0].timestamp

            for (e in original) {
                if (e.timestamp < from) from = e.timestamp
                if (e.timestamp > to) to = e.timestamp
            }
            for (interval in intervals) {
                if (interval.begin < from) from = interval.begin
                if (interval.end > to) to = interval.end
            }

            // Create unknown entries
            var current = to
            while (current >= from) {
                result.add(Entry(current, UNKNOWN))
                current = current.minus(1)
            }

            // Create YES_AUTO entries
            intervals.forEach { interval ->
                current = interval.end
                while (current >= interval.begin) {
                    val offset = current.daysUntil(to)
                    result[offset] = Entry(current, YES_AUTO)
                    current = current.minus(1)
                }
            }

            // Copy original entries
            original.forEach { entry ->
                val offset = entry.timestamp.daysUntil(to)
                if (result[offset].value == UNKNOWN || entry.value == SKIP || entry.value == YES_MANUAL) {
                    result[offset] = entry
                }
            }

            return result
        }

        /**
         * Starting from the second newest interval, this function tries to slide the
         * intervals backwards into the past, so that gaps are eliminated and
         * streaks are maximized.
         *
         * The intervals should be sorted by timestamp. The first element in the list should
         * correspond to the newest interval.
         */
        fun snapIntervalsTogether(intervals: ArrayList<Interval>) {
            for (i in 1 until intervals.size) {
                val curr = intervals[i]
                val next = intervals[i - 1]
                val gapNextToCurrent = next.begin.daysUntil(curr.end)
                val gapCenterToEnd = curr.center.daysUntil(curr.end)
                if (gapNextToCurrent >= 0) {
                    val shift = min(gapCenterToEnd, gapNextToCurrent + 1)
                    intervals[i] = Interval(
                        curr.begin.minus(shift),
                        curr.center,
                        curr.end.minus(shift)
                    )
                }
            }
        }

        fun buildIntervals(
            freq: Frequency,
            entries: List<Entry>,
        ): ArrayList<Interval> {
            val filtered = entries.filter { it.value == YES_MANUAL }
            val num = freq.numerator
            val den = freq.denominator
            val intervals = arrayListOf<Interval>()
            for (i in num - 1 until filtered.size) {
                val (begin, _) = filtered[i]
                val (center, _) = filtered[i - num + 1]
                if (begin.daysUntil(center) < den) {
                    val end = begin.plus(den - 1)
                    intervals.add(Interval(begin, center, end))
                }
            }
            return intervals
        }
    }
}
